### PR TITLE
Select NTLM9 password characters with div/mod, rather than more biase…

### LINKED ISSUE #2: Bias in index_to_plaintext_ntlm9()
As mentioned in #1 and [this Reddit comment](https://www.reddit.com/r/netsec/comments/bzae2t/rainbow_crackalack_make_rainbow_tables_great_again/equfzie?utm_source=share&utm_medium=web2x), there is a bias in the implementation of the `index_to_plaintext_ntlm9()` function in `CL/ntlm9_functions.cl`.

### PR DESCRIPTION
**Pull request by pruby**
…d and/mod/shift.

This still leaves a small bias towards a part of the password set. There are 29 complete
repetitions of all possible passwords within the 64-bit index space, followed by about a
quarter cycle which must be somehow biased as it can't cover everything.

This bias could be removed/reduced using larger indices, however doing so may be a poor
tradeoff. Assuming pessimistically that this is "wasted" would represent 0.92% of chain
entries being a waste. Because some work is repeated, these could lead to a 1.83%
performance loss at a chain length of 1,350,000. If avoiding the bias by using a larger
index, we would want to see no more than a 1.8% performance reduction to do so. Fast
and simple would be my choice in this case.

Fixes #2

### CODE REVIEW

**Comment by pruby** [comment]
If interested in how I calculate that 1.83% loss, it's because to test a rainbow table with chain length X takes X*X/2 operations. If Y entries have zero value, then the performance of the chain is equivalent to a shorter chain of length (X-Y) with no duds. The work that would be required to test the chain without duds is therefore (X-Y)*(X-Y)/2.

The loss can then be calculated as 1 - ((X-Y)*(X-Y) / (X*X)). With X=1,350,000 and Y=(0.0092 * 1,350,000) this produces just under 1.83%.

Of course, this is pessimistic - the biased portion is just _less_ useful, not useless.

**Comment by pruby** [comment]
OK, so there are some real oddities in unit tests with that last change... trying to work out why the CPU and GPU produce different results after that change and the same result before (despite the GPU code being changed!)

**Comment by pruby** [comment]
Unit tests fixed. Spent way more time than expected on a quick fix, but that should do it.

File: CL/ntlm9_functions.cl [hunk 1]
```diff
@@ -11,8 +11,8 @@ __constant char charset[] = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQ
 inline void index_to_plaintext_ntlm9(unsigned long index, __constant char *charset, unsigned char *plaintext) {
 
   for (int i = 0; i < 9; i++) {
```
Comment (Sc00bz) [inline CL/ntlm9_functions.cl]: Should be like this to match CPU and rcrack:
for (int i = 8; i >= 0; i--) {

**Comment by pruby** [comment]
True. Turns out by the way that the optimised kernel here is untested, hence not catching things like this.

**Comment by pruby** [comment]
Got a bug trying to build a table matching this branch, should be completely unrelated though.

Table is in Amazon S3 (should be public, but need an AWS account to download):

ml-data.goddard.nz/public/ntlm_ascii-32-95%239-9_0_1350000x67108864_0.rt

Log is in the same place:

ml-data.goddard.nz/public/ntlm_ascii-32-95%239-9_0_1350000x67108864_0.rt.log

Errors shown were:

```
Estimated time remaining: 13 mins, 18 secs                                                                           [61/1976]
Run time: 3 hours, 44 mins; Chains generated: 62791680; Rate: 4651/s
Estimated time remaining: 12 mins, 4 secs
Run time: 3 hours, 46 mins; Chains generated: 63157248; Rate: 4653/s
Estimated time remaining: 10 mins, 45 secs
Run time: 3 hours, 47 mins; Chains generated: 63501312; Rate: 4653/s
Estimated time remaining: 9 mins, 31 secs
Run time: 3 hours, 48 mins; Chains generated: 63845376; Rate: 4653/s
Estimated time remaining: 8 mins, 17 secs
Run time: 3 hours, 49 mins; Chains generated: 64189440; Rate: 4653/s
Estimated time remaining: 7 mins, 4 secs
Run time: 3 hours, 51 mins; Chains generated: 64533504; Rate: 4653/s
Estimated time remaining: 5 mins, 50 secs
Run time: 3 hours, 52 mins; Chains generated: 64856064; Rate: 4652/s
Estimated time remaining: 4 mins, 40 secs
Run time: 3 hours, 53 mins; Chains generated: 65200128; Rate: 4652/s
Estimated time remaining: 3 mins, 26 secs
Run time: 3 hours, 54 mins; Chains generated: 65544192; Rate: 4652/s
Estimated time remaining: 2 mins, 12 secs
Run time: 3 hours, 56 mins; Chains generated: 65888256; Rate: 4652/s
Estimated time remaining: 59.0 secs
Run time: 3 hours, 57 mins; Chains generated: 66167808; Rate: 4648/s
Estimated time remaining: 47457 days, 16 hours
Run time: 3 hours, 57 mins; Chains generated: 66167808; Rate: 4648/s

Generation complete!
Now verifying rainbow table... Error: table is expected to be complete, but file size does not match expected value.  Expected
: 1073741824; actual: 1058684928

Error while verifying rainbowtable!  It has been truncated to just before the point of error.  Please give the following file 
to the developer: ntlm_ascii-32-95#9-9_0_1350000x67108864_0.rt.log
```

**Pull request merged**